### PR TITLE
Use DCP in HFCheckpointer to read/write directly to HuggingFace

### PR DESCRIPTION
**Pull request by ankitageorge**
#### Context
What is the purpose of this PR? Is it to
- [ x] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

Please link to any issues this PR addresses.

#### Changelog
What are the changes made in this PR?
* Add the ability to read and write checkpoints directly in HF format to HF, behind the enable_dcp flag. 

#### Test plan
Please make sure to do each of the following if applicable to your PR. If you're unsure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.

- [ x] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [ x] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [ x] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [x ] run unit tests via `pytest tests`
- [ x] run recipe tests via `pytest tests -m integration_test`
- [ x] manually run any new or modified recipes with sufficient proof of correctness
- [ x] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Here is a [docstring example](https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285)
and a [tutorial example](https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models)

- [ ] I did not change any public API
- [x ] I have added an example to docs or docstrings

### CODE REVIEW

File: torchtune/training/checkpointing/_utils.py [hunk 1]
```diff
@@ -581,15 +600,21 @@ def validate_checkpoint_files(
...
+) -> bool:
     """
     Checks that the output directory is not equal to or a subdirectory of the checkpoint directory.
     This is necessary to avoid making copies of copies when geting config files from ckpt_dir.
     """
-
     # Resolve the absolute paths to avoid issues with relative paths
-    _ckpt_dir = ckpt_dir.resolve()
-    _out_dir = out_dir.resolve()
+    if isinstance(ckpt_dir, Path):
```
Comment (joecummings) [inline torchtune/training/checkpointing/_utils.py:611]: Should probably just convert to path to start and then resolve.

File: torchtune/training/checkpointing/_utils.py [hunk 2]
```diff
@@ -236,22 +241,31 @@ def safe_torch_load(
...
-                weights_only=weights_only,
-            )
+            if isinstance(fs, HfFileSystem):
+                # HfFileSystem does not support mmap
+                mmap = False
+                with fs.open(checkpoint_path, "rb") as checkpoint_file:
+                    state_dict = torch.load(
+                        checkpoint_file,
+                        map_location="cpu",
+                        mmap=mmap,
```
Comment (joecummings) [inline torchtune/training/checkpointing/_utils.py:259]: Will this be much slower without mmap then?

**Review by joecummings — COMMENTED** [review]
Very small comments - would you mind just manually verifying that you can download directly from a HF Hub repo and upload to a new Repo as a test?

File: torchtune/training/checkpointing/_utils.py (hunk 1 above)
Comment (ankitageorge) [reply torchtune/training/checkpointing/_utils.py:611]: for remote paths this won't work. they don't have a resolve method, that's why I did this. if it's a remote path like hf, it would be a string and not call resolve

File: torchtune/training/checkpointing/_utils.py (hunk 2 above)
Comment (ankitageorge) [reply torchtune/training/checkpointing/_utils.py:259]: yes, for large files. But this code path would be for the extra files that don't go through dcp like, adapter checkpoint and recipe checkpoint which would be smaller files right?

**Comment by ankitageorge** [comment]
> Very small comments - would you mind just manually verifying that you can download directly from a HF Hub repo and upload to a new Repo as a test?

Yes, this is working
